server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -527,6 +531,128 @@ private void checkAndAddInEligibleTargetNode(RoutingNode targetNode) {
         }
     }
 
+    /**
+     * Triggers shard relocation for the shards that are submitted
+     * for tiering to go from local node pool to remote capable node
+     * pool.
+     */
+    public void tierShards() {
+        List<ShardRouting> shardsPendingTiering = new ArrayList<>();
+        List<RoutingNode> remoteRoutingNodes = new ArrayList<>();
+        Iterator<RoutingNode> nodes = routingNodes.mutableIterator();
+        while (nodes.hasNext()) {
+            RoutingNode node = nodes.next();
+            RoutingPool pool = RoutingPool.getNodePool(node);
+            if (pool == RoutingPool.REMOTE_CAPABLE) {
+                remoteRoutingNodes.add(node);
+            }
+        }
+
+        /* Routing Table is created after reroute and is immutable. It does not hold
+         * updated shard states with changes applied during the reroute cycle. Throughout the
+         * allocation (reroute) cycle, we need to work on RoutingNodes shard objects.
+         */
+        List<ShardRouting> indexShards = allocation.routingTable().allShards();
+        for (ShardRouting shard : indexShards) {
```

Comment:
yeah this approach was discussed in https://github.com/opensearch-project/OpenSearch/issues/13294, however adding the tiering entry in cluster state would impact the number of cluster state updates since we support tiering of multiple indices in a request which would be expensive. This operation here in re-route would still be O(number of shards). Thoughts?

Reply:
Can we simply this or add caching, iterating over large number of shards is not dirt cheap

Reply:
makes sense, thinking of simplifying it as :

```
final Set<String> indicesPendingTiering = new HashSet<>();
allocation.routingTable().indicesRouting().keySet().stream().parallel().forEach(index -> {
    IndexMetadata indexMetadata = allocation.metadata().index(index);
    if (IndexModule.TieringState.HOT_TO_WARM.name()
        .equals(indexMetadata.getSettings().get(IndexModule.INDEX_TIERING_STATE.getKey()))) {
        indicesPendingTiering.add(index);
    }
});
if (indicesPendingTiering.isEmpty()) {
    logger.debug("No indices found eligible for hot to warm tiering.");
    return;
}
logger.debug("Indices pending hot to warm tiering are [{}]", indicesPendingTiering);
final List<ShardRouting> shardsPendingTiering = new ArrayList<>();
for (String index : indicesPendingTiering) {
    final List<ShardRouting> indexShards = allocation.routingTable().shardsWithState(ShardRoutingState.STARTED);
    logger.debug("Found shards {} in STARTED state for index [{}] for tiering from hot to warm", indexShards, index);
    for (ShardRouting shard : indexShards) {
        RoutingPool targetPool = RoutingPool.getShardPool(shard, allocation);
        RoutingPool currentNodePool = RoutingPool.getNodePool(allocation.routingNodes().node(shard.currentNodeId()));
        if (targetPool != currentNodePool) {
            logger.trace(
                "Found shard [{}] with target pool [{}] on node pool [{}]. Adding to tiering list.",
                shard.toString(),
                targetPool.toString(),
                currentNodePool.toString()
            );
            shardsPendingTiering.add(shard);
        }
    }
}
```

@kotwanikunal @Bukhtawar let me know how this looks to you